### PR TITLE
Refactor java array accessors

### PR DESCRIPTION
**Pull request by beeender**
The original implementation has a serious memory issue that the
ReleaseXxxArrayElements might be called before accessing if the
copy/move constructor is invoked. That could happen when use
JniArraysOfArray where the JniXxxArray would be copied when push it to
the vector. The JNI SortDescriptor may trigger various native crashes
because of using this.

### CODE REVIEW

File: realm/realm-library/src/main/cpp/jni_util/java_exception_thrower.hpp [hunk 1]
```diff
@@ -21,6 +21,8 @@
 
 #include <stdexcept>
 
+#include <util/format.hpp>
```
Comment (cmelchior) [inline realm/realm-library/src/main/cpp/jni_util/java_exception_thrower.hpp]: Doesn't look to be used?

**Comment by beeender** [comment]
Yes! easy to expand to support that!

File: realm/realm-library/src/main/cpp/java_accessor.hpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        return m_elements_holder->m_data_ptr;
+    }
+
+    inline const ElementType& operator[](const int index) const noexcept
+    {
+        return m_elements_holder->m_data_ptr[index];
+    }
+
+    template <typename T>
+    T transform();
```
Comment (cmelchior) [inline realm/realm-library/src/main/cpp/java_accessor.hpp]: I would add a comment here about what what was tranformed to what.

File: realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp [hunk 3]
```diff
@@ -305,30 +308,17 @@ JNIEXPORT void JNICALL Java_io_realm_internal_UncheckedRow_nativeSetByteArray(JN
...
-            if (!(ROW(nativeRowPtr)->get_table()->is_nullable(S(columnIndex)))) {
-                ThrowNullValueException(env, ROW(nativeRowPtr)->get_table(), S(columnIndex));
-                return;
-            }
-            ROW(nativeRowPtr)->set_binary(S(columnIndex), BinaryData());
-        }
-        else {
-            bytePtr = env->GetByteArrayElements(value, NULL);
-            if (!bytePtr) {
-                ThrowException(env, IllegalArgument, "doByteArray");
```
[the remark this answers is not in the document]
Comment (beeender) [reply realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp]: yes! the original exception message is funny "doByteArray"

File: realm/realm-library/src/main/cpp/java_accessor.hpp [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+template <>
+inline JPrimitiveArrayAccessor<jbyteArray, jbyte>::ElementsHolder::~ElementsHolder()
+{
+    if (m_jarray) {
+        m_env->ReleaseByteArrayElements(m_jarray, m_data_ptr, m_release_mode);
+    }
+}
+
+template <>
```
[the remark this answers is not in the document]
Comment (beeender) [reply realm/realm-library/src/main/cpp/java_accessor.hpp]: https://stackoverflow.com/questions/8640390/template-function-inside-template-class all magic!

File: realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kneth) [reply realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp]: It dates back to a time where we experimented with different methods to copy byte array as fast as possible.

**Pull request merged**